Mathlib/Lean/Expr/ReplaceRec.lean

```diff
@@ -0,0 +1,53 @@
+/-
+Copyright (c) 2019 Robert Y. Lewis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Simon Hudon, Scott Morrison, Keeley Hoek, Robert Y. Lewis, Floris van Doorn, E.W.Ayers
+-/
+import Lean
+import Mathlib.Util.TermUnsafe
+import Mathlib.Lean.Expr.Traverse
+import Mathlib.Util.MemoFix
+namespace Lean.Expr
+/-!
+# ReplaceRec
+
+We define a more flexible version of `Expr.replace` where we can use recursive calls even when
+replacing a subexpression. We completely mimic the implementation of `Expr.replace`. -/
+
+/-- Same as `replaceRec` except over a monad. -/
+def replaceRecM [Monad M] (f? : (Expr → M Expr) → Expr → M (Option Expr)) : Expr → M Expr :=
```

Comment:
I didn't realize this until now, but `replaceRecM` won't do what you expect it to.  It will cache the `M Expr` and not the `Expr`, so for most monads this won't do any caching at all.  `replaceRec` is fine though.

Reply:
Yeah that's what I thought it was doing and I agree it's pointless

Reply:
but it seemed neater to implement like this.

There is also a method in `src/lean/meta/transform.lean` that does the same kind of thing as this. Maybe this whole PR could be replaced with a call to that.

Reply:
I think that there might be a way of implementing proper monad caching using `[MonadControlT _ M]` but I'm not sure.

Reply:
In their current state, I'd prefer to remove the `replaceRecM` and co., since they have very misleading performance (and thus should not be used).

To properly implement them, I think we'd need a `memoFixM` combinator.

The `Meta.transform` function is a tradeoff.  It is nicer since it changes the local context, which means you can call `inferType` etc. on the subexpression.  But you can't perform any recursive calls, just change the expression at the current position.  (Which, to be fair, should be enough for `to_additive`.)